File: wrappers/typescript/src/l14.ts
/**
 * L14 safety net — client-side result/final synthesis (design §4.6 contract #1).
 *
 * If the engine emits a non-null reply in its turn/submit response BUT no
 * result/final notification was observed first, the wrapper synthesizes a
 * result/final-shaped DisplayEvent with synthesized: true.
 *
 * Branch A (engine emits result/final) → sawFinal=true → no synthesis.
 * Branch B (engine omits result/final) → sawFinal=false, reply≠null → synthesize.
 */

import type { DisplayEvent } from "./session.js";

/** Parameters for synthesizeFinalIfMissing. */
export interface SynthesizeParams {
  /** True if a result/final notification was already observed during the turn. */
  sawFinal: boolean;
  /** The reply text from the turn/submit response, or null if absent. */
  reply: string | null;
  sessionId: string;
  turnId: string;
}

/**
 * Synthesize a result/final DisplayEvent if the engine omitted it.
 *
 * Returns null if:
 * - sawFinal is true (result/final was already observed), OR
 * - reply is null (no reply to synthesize from)
 *
 * Otherwise returns a DisplayEvent with type='result/final' and synthesized=true.
 */
export function synthesizeFinalIfMissing({
  sawFinal,
  reply,
  sessionId,
  turnId,
}: SynthesizeParams): DisplayEvent | null {
  if (sawFinal || reply === null) {
    return null;
  }
  return {
    type: "result/final",
    sessionId,
    turnId,
    synthesized: true,
    payload: {
      sessionId,
      turnId,
      text: reply,
    },
  };
}

File: wrappers/python/src/amplifier_agent_client/l14.py
"""L14 safety net — client-side result/final synthesis (design §4.6 contract #1).

If the engine emits a non-null reply in its turn/submit response BUT no
result/final notification was observed first, the wrapper synthesizes a
result/final-shaped DisplayEvent dict with synthesized=True.

Branch A (engine emits result/final) → saw_final=True → no synthesis.
Branch B (engine omits result/final) → saw_final=False, reply≠None → synthesize.
"""

from __future__ import annotations

from typing import Any


def synthesize_final_if_missing(
    *,
    saw_final: bool,
    reply: str | None,
    session_id: str,
    turn_id: str,
) -> dict[str, Any] | None:
    """Synthesize a result/final DisplayEvent dict if the engine omitted it.

    Returns None if:
    - saw_final is True (result/final was already observed), OR
    - reply is None (no reply to synthesize from)

    Otherwise returns a dict shaped like a result/final DisplayEvent with
    synthesized=True.
    """
    if saw_final or reply is None:
        return None
    return {
        "type": "result/final",
        "session_id": session_id,
        "turn_id": turn_id,
        "synthesized": True,
        "payload": {
            "sessionId": session_id,
            "turnId": turn_id,
            "text": reply,
        },
    }
